File: 3753-total-waviness-of-numbers-in-range-ii/3753-total-waviness-of-numbers-in-range-ii.java
class Solution {
    public long totalWaviness(long num1, long num2) {
        return countUpTo(num2) - countUpTo(num1 - 1);
    }

    private long countUpTo(long num) {
        if (num < 100) {
            return 0;
        }

        String s = Long.toString(num);
        int n = s.length();

        // dp[idx][prev1][prev2][isLess][isStart]
        // prev1, prev2는 0~9(미지정 시 10) 총 11가지 상태 필요
        long[][][][][] dp = new long[n][11][11][2][2];
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < 11; j++) {
                for (int k = 0; k < 11; k++) {
                    for (int l = 0; l < 2; l++) {
                        Arrays.fill(dp[i][j][k][l], -1);
                    }
                }
            }
        }

        return dfs(0, 10, 10, false, true, s, dp);
    }

    private long dfs(int idx, int prev1, int prev2, boolean isLess, boolean isStart, String s, long[][][][][] dp) {
        int n = s.length();
        if (idx == n) {
            return 0;
        }

        int il = isLess ? 1 : 0;
        int is = isStart ? 1 : 0;

        if (dp[idx][prev1][prev2][il][is] != -1) {
            return dp[idx][prev1][prev2][il][is];
        }

        int limit = isLess ? 9 : (s.charAt(idx) - '0');
        long res = 0;

        for (int i = 0; i <= limit; i++) {
            boolean nextLess = isLess || (i < limit);
            boolean nextStart = isStart && (i == 0);

            // 피크(Peak) 및 밸리(Valley) 체크
            long waviness = 0;
            if (!isStart && prev1 != 10 && prev2 != 10) {
                if (prev1 > prev2 && prev1 > i) { // Peak
                    waviness = 1;
                } else if (prev1 < prev2 && prev1 < i) { // Valley
                    waviness = 1;
                }
            }

            int n1, n2;
            if (nextStart) {
                n1 = 10;
                n2 = 10;
            } else {
                n1 = i;
                n2 = prev1;
            }

            // 현재 자릿수의 waviness 조합 수 계산 후 재귀 호출 누적
            long combinations = countCombinations(n, idx + 1, nextLess, s);
            res += waviness * combinations + dfs(idx + 1, n1, n2, nextLess, nextStart, s, dp);
        }

        return dp[idx][prev1][prev2][il][is] = res;
    }

    // 이후 자릿수에서 발생할 수 있는 유효 숫자 조합의 수를 반환하는 보조 함수
    private long countCombinations(int length, int idx, boolean isLess, String s) {
        if (idx == length) {
            return 1;
        }
        if (isLess) {
            long res = 1;
            for (int i = idx; i < length; i++) {
                res *= 10;
            }
            return res;
        }
        return Long.parseLong(s.substring(idx)) + 1;
    }
}

File: 3753-total-waviness-of-numbers-in-range-ii/3753-total-waviness-of-numbers-in-range-ii.ts
function totalWaviness(num1: number, num2: number): number {
    // JS/TS 안전 정수 범위를 벗어날 수 있으므로 계산은 BigInt로 진행합니다.
    const b1 = BigInt(num1);
    const b2 = BigInt(num2);
    
    return Number(countUpTo(b2) - countUpTo(b1 - 1n));
}

function countUpTo(num: bigint): bigint {
    if (num < 100n) {
        return 0n;
    }

    const s = num.toString();
    const n = s.length;

    // DP 상태 개수 최적화 및 1차원 평탄화 배열 크기 계산
    // idx(최대 16) * prev1(11) * prev2(11) * isLess(2) * isStart(2)
    const STATE_SIZE = n * 11 * 11 * 2 * 2;
    const dp = new BigInt64Array(STATE_SIZE).fill(-1n);

    // 5차원 인덱스를 1차원 배열 인덱스로 매핑해주는 헬퍼 함수
    function getStateIndex(idx: number, prev1: number, prev2: number, isLess: number, isStart: number): number {
        return ((((idx * 11 + prev1) * 11 + prev2) * 2 + isLess) * 2 + isStart);
    }

    function dfs(idx: number, prev1: number, prev2: number, isLess: boolean, isStart: boolean): bigint {
        if (idx === n) {
            return 0n;
        }

        const il = isLess ? 1 : 0;
        const is = isStart ? 1 : 0;
        const stateIdx = getStateIndex(idx, prev1, prev2, il, is);

        if (dp[stateIdx] !== -1n) {
            return dp[stateIdx];
        }

        const limit = isLess ? 9 : parseInt(s[idx], 10);
        let res = 0n;

        for (let i = 0; i <= limit; i++) {
            const nextLess = isLess || i < limit;
            const nextStart = isStart && i === 0;

            // 피크(Peak) 및 밸리(Valley) 체크 조건식
            let waviness = 0n;
            if (!isStart && prev1 !== 10 && prev2 !== 10) {
                if (prev1 > prev2 && prev1 > i) {       // Peak
                    waviness = 1n;
                } else if (prev1 < prev2 && prev1 < i) { // Valley
                    waviness = 1n;
                }
            }

            let n1 = 10, n2 = 10;
            if (!nextStart) {
                n1 = i;
                n2 = prev1;
            }

            // 현재 자릿수의 유효 숫자 조합 수 계산 후 재귀 호출 누적
            const combinations = countCombinations(n, idx + 1, nextLess, s);
            res += waviness * combinations + dfs(idx + 1, n1, n2, nextLess, nextStart);
        }

        dp[stateIdx] = res;
        return res;
    }

    return dfs(0, 10, 10, false, true);
}

// 해당 자릿수 뒤로 나타날 수 있는 유효 숫자 조합의 총 개수를 구하는 보조 함수
function countCombinations(length: number, idx: number, isLess: boolean, s: string): bigint {
    if (idx === length) {
        return 1n;
    }
    if (isLess) {
        let res = 1n;
        for (let i = idx; i < length; i++) {
            res *= 10n;
        }
        return res;
    }
    return BigInt(s.substring(idx)) + 1n;
}
